{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Création des fichiers explicites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A ne pas refaire tourner"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On reprend les bases de données auxquelles on ajoute une colonne détaillant les séquences concernées, qu'on extrait du génome explicite."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Base de donnée des G4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_chr = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y']\n",
    "list = []\n",
    "\n",
    "file = open('C:/Users/BlueC/Documents/Mines/Santé/Code/annotated_G4plus_categories.bed')\n",
    "reader = pd.read_csv(file, delimiter='\\t')\n",
    "\n",
    "col = reader.columns.tolist()\n",
    "col1 = reader.columns.tolist()\n",
    "col.append('sequence')\n",
    "\n",
    "file_f = pd.DataFrame(columns = col)\n",
    "file_f[col1] = reader\n",
    "\n",
    "for i in list_chr:\n",
    "\n",
    "    path = f'C:/Users/BlueC/Documents/Mines/Santé/Code/human_genome/chr{i}_G4_explicit'\n",
    "\n",
    "    with open(path) as file:\n",
    "        for line in file:\n",
    "            line1 = line\n",
    "            line2 = next(file)\n",
    "            stripped_line = line2.strip()\n",
    "            list.append(stripped_line)\n",
    "\n",
    "\n",
    "    print(i)\n",
    "\n",
    "file_f['sequence'] = list\n",
    "file_f.to_csv('C:/Users/BlueC/Documents/Mines/Santé/Code/annotated_G4plus_categories_explicit.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Positions random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_chr = [1,2,3,4,5,6,7,'X',8,9,10,11,12,13,14,15,16,17,18,20,'Y',19,22,21]\n",
    "list = []\n",
    "\n",
    "file_g = open('C:/Users/BlueC/Documents/Mines/Santé/Code/random_G4.csv')\n",
    "reader = pd.read_csv(file_g, delimiter=',')\n",
    "\n",
    "reader = reader.drop('Unnamed: 0', axis = 1)\n",
    "\n",
    "col = reader.columns.tolist()\n",
    "col1 = reader.columns.tolist()\n",
    "col.append('sequence')\n",
    "\n",
    "file_f = pd.DataFrame(columns = col)\n",
    "file_f[col1] = reader\n",
    "\n",
    "for i in list_chr:\n",
    "\n",
    "    path = f'C:/Users/BlueC/Documents/Mines/Santé/Code/human_genome_random/chr{i}_G4_random_explicit'\n",
    "\n",
    "    with open(path) as file:\n",
    "        for line in file:\n",
    "            line1 = line\n",
    "            line2 = next(file)\n",
    "            stripped_line = line2.strip()\n",
    "            list.append(stripped_line)\n",
    "\n",
    "\n",
    "    print(i)\n",
    "\n",
    "file_f['sequence'] = list\n",
    "file_f.to_csv('C:/Users/BlueC/Documents/Mines/Santé/Code/random_G4_explicit.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open('C:/Users/BlueC/Documents/Mines/Santé/Code/annotated_G4plus_categories_explicit.csv')\n",
    "reader = pd.read_csv(file, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>distClosestG4</th>\n",
       "      <th>CGI</th>\n",
       "      <th>Ori</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28893</td>\n",
       "      <td>28915</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>134</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGACCAGGAGCTCCGGGCGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29048</td>\n",
       "      <td>29073</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>134</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGCCGGGCAAGGCCGGGTGCAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>32088</td>\n",
       "      <td>32114</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>1156</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGCCAGCTGGGCTGAGCGGGCCTGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33269</td>\n",
       "      <td>33294</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>1076</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ggggctaaggggaggggagaatggg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>34369</td>\n",
       "      <td>34388</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>634</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGGAGGGTGGGCAAAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35021</td>\n",
       "      <td>35039</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>634</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGAGAGGGGTGGAGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36080</td>\n",
       "      <td>36104</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>154</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGAGCTGGGCACAGGGCAGCAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36257</td>\n",
       "      <td>36289</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>154</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGTGGCTCTGAGGGGGCTCTCAGGGGTGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36848</td>\n",
       "      <td>36871</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGTGCTGGCGATGGGGTGGGGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr  start    end       type  distClosestG4  CGI  Ori  TSS  \\\n",
       "16  chr1  28893  28915      Bulge            134    1    0    0   \n",
       "17  chr1  29048  29073  Canonical            134    1    0    0   \n",
       "18  chr1  32088  32114       GVBQ           1156    0    0    0   \n",
       "19  chr1  33269  33294  Canonical           1076    0    0    0   \n",
       "20  chr1  34369  34388  Canonical            634    0    0    0   \n",
       "21  chr1  35021  35039       GVBQ            634    0    0    0   \n",
       "22  chr1  36080  36104       GVBQ            154    0    0    0   \n",
       "23  chr1  36257  36289      Bulge            154    0    0    0   \n",
       "24  chr1  36848  36871      Bulge            560    0    0    0   \n",
       "\n",
       "                            sequence  \n",
       "16            GGGACCAGGAGCTCCGGGCGGG  \n",
       "17         GGGCCGGGCAAGGCCGGGTGCAGGG  \n",
       "18        GGCCAGCTGGGCTGAGCGGGCCTGGG  \n",
       "19         ggggctaaggggaggggagaatggg  \n",
       "20               GGGGAGGGTGGGCAAAGGG  \n",
       "21                GGGAGAGGGGTGGAGGGG  \n",
       "22          GGGAGCTGGGCACAGGGCAGCAGG  \n",
       "23  GGGTGGCTCTGAGGGGGCTCTCAGGGGTGGGG  \n",
       "24           GGGTGCTGGCGATGGGGTGGGGG  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reader[16:25]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Création des datasets de mutations - G4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Création des fichiers textes de comparaison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A ne pas refaire tourner"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On crée des fichiers textes dans lequels sont recensés le numéro du chromosome concerné, ainsi que toutes les positions de début et de fin des G4. Ces fichiers serviront ensuite à extraire les mutations ayant lieu dans lesdites zones, grâce à un code tournant en bash sur serveur."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range (1,23):\n",
    "    test = reader[(reader['chr'] == f'chr{i}')]\n",
    "    test = test[['chr','start','end']]\n",
    "    test = test.sort_values(['start','end'])\n",
    "    test['chr'] = i\n",
    "    path = f\"c:/Users/BlueC/Documents/Mines/Santé/Code/np_{i}.txt\"\n",
    "    np.savetxt(path, test.values, fmt='%s', delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in (['X','Y']):\n",
    "    test = reader[(reader['chr'] == f'chr{i}')]\n",
    "    test = test[['chr','start','end']]\n",
    "    test = test.sort_values(['start','end'])\n",
    "    test['chr'] = i\n",
    "    path = f\"c:/Users/BlueC/Documents/Mines/Santé/Code/np_{i}.txt\"\n",
    "    np.savetxt(path, test.values, fmt='%s', delimiter='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Création de la base recensant toutes les mutations ayant eu lieu sur les G4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On récupère les fichiers générés grâce au code bash précédemment évoqué pour créer une base de données complète, recensant les mutations et les caractéristiques du G4 auquel elles sont associées."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open('C:/Users/BlueC/Documents/Mines/Santé/Code/annotated_G4plus_categories.bed')\n",
    "reader = pd.read_csv(file, delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "i_chrom = '1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>name</th>\n",
       "      <th>orig</th>\n",
       "      <th>mut</th>\n",
       "      <th>%</th>\n",
       "      <th>PASS</th>\n",
       "      <th>INFOS</th>\n",
       "      <th>AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11008</td>\n",
       "      <td>rs575272151</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=441;AF=0.0880591;AN=5008;NS=2504;DP=2232;EA...</td>\n",
       "      <td>.|||;VT=SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>11012</td>\n",
       "      <td>rs544419019</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=441;AF=0.0880591;AN=5008;NS=2504;DP=2090;EA...</td>\n",
       "      <td>.|||;VT=SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>11063</td>\n",
       "      <td>rs561109771</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=15;AF=0.00299521;AN=5008;NS=2504;DP=2834;EA...</td>\n",
       "      <td>.|||;VT=SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14674</td>\n",
       "      <td>rs561913721</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;AF=0.000199681;AN=5008;NS=2504;DP=26402;E...</td>\n",
       "      <td>g|||;VT=SNP;EX_TARGET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16738</td>\n",
       "      <td>rs549612027</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;AF=0.000199681;AN=5008;NS=2504;DP=16499;E...</td>\n",
       "      <td>g|||;VT=SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65961</th>\n",
       "      <td>1</td>\n",
       "      <td>249240217</td>\n",
       "      <td>rs570135855;rs556759920</td>\n",
       "      <td>G</td>\n",
       "      <td>GT,T</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=368,183;AF=0.0734824,0.0365415;AN=5008;NS=2...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65962</th>\n",
       "      <td>1</td>\n",
       "      <td>249240219</td>\n",
       "      <td>rs370070114</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1454;AF=0.290335;AN=5008;NS=2504;DP=149437;...</td>\n",
       "      <td>.|||;VT=SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65963</th>\n",
       "      <td>1</td>\n",
       "      <td>249240537</td>\n",
       "      <td>rs549505823</td>\n",
       "      <td>GGT</td>\n",
       "      <td>G</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=2272;AF=0.453674;AN=5008;NS=2504;DP=66563;E...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65964</th>\n",
       "      <td>1</td>\n",
       "      <td>249240539</td>\n",
       "      <td>rs545682552</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1682;AF=0.335863;AN=5008;NS=2504;DP=56555;E...</td>\n",
       "      <td>.|||;VT=SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65965</th>\n",
       "      <td>1</td>\n",
       "      <td>249240543</td>\n",
       "      <td>rs536805817</td>\n",
       "      <td>AGG</td>\n",
       "      <td>A</td>\n",
       "      <td>100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=3098;AF=0.61861;AN=5008;NS=2504;DP=69629;EA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65966 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr        pos                     name orig   mut    %  PASS  \\\n",
       "0        1      11008              rs575272151    C     G  100  PASS   \n",
       "1        1      11012              rs544419019    C     G  100  PASS   \n",
       "2        1      11063              rs561109771    T     G  100  PASS   \n",
       "3        1      14674              rs561913721    G     A  100  PASS   \n",
       "4        1      16738              rs549612027    G     A  100  PASS   \n",
       "...    ...        ...                      ...  ...   ...  ...   ...   \n",
       "65961    1  249240217  rs570135855;rs556759920    G  GT,T  100  PASS   \n",
       "65962    1  249240219              rs370070114    A     T  100  PASS   \n",
       "65963    1  249240537              rs549505823  GGT     G  100  PASS   \n",
       "65964    1  249240539              rs545682552    T     G  100  PASS   \n",
       "65965    1  249240543              rs536805817  AGG     A  100  PASS   \n",
       "\n",
       "                                                   INFOS  \\\n",
       "0      AC=441;AF=0.0880591;AN=5008;NS=2504;DP=2232;EA...   \n",
       "1      AC=441;AF=0.0880591;AN=5008;NS=2504;DP=2090;EA...   \n",
       "2      AC=15;AF=0.00299521;AN=5008;NS=2504;DP=2834;EA...   \n",
       "3      AC=1;AF=0.000199681;AN=5008;NS=2504;DP=26402;E...   \n",
       "4      AC=1;AF=0.000199681;AN=5008;NS=2504;DP=16499;E...   \n",
       "...                                                  ...   \n",
       "65961  AC=368,183;AF=0.0734824,0.0365415;AN=5008;NS=2...   \n",
       "65962  AC=1454;AF=0.290335;AN=5008;NS=2504;DP=149437;...   \n",
       "65963  AC=2272;AF=0.453674;AN=5008;NS=2504;DP=66563;E...   \n",
       "65964  AC=1682;AF=0.335863;AN=5008;NS=2504;DP=56555;E...   \n",
       "65965  AC=3098;AF=0.61861;AN=5008;NS=2504;DP=69629;EA...   \n",
       "\n",
       "                          AA  \n",
       "0                .|||;VT=SNP  \n",
       "1                .|||;VT=SNP  \n",
       "2                .|||;VT=SNP  \n",
       "3      g|||;VT=SNP;EX_TARGET  \n",
       "4                g|||;VT=SNP  \n",
       "...                      ...  \n",
       "65961                   None  \n",
       "65962            .|||;VT=SNP  \n",
       "65963                   None  \n",
       "65964            .|||;VT=SNP  \n",
       "65965                   None  \n",
       "\n",
       "[65966 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file2 = open(f'C:/Users/BlueC/Documents/Mines/Santé/Code/chr/chr_{i_chrom}_2')\n",
    "my_col = ['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA']\n",
    "separateurs = \"\\t|AA=\"\n",
    "reader2 = pd.read_csv(file2, sep = separateurs, names = my_col, header = None, engine ='python')\n",
    "reader2 = reader2.drop_duplicates()\n",
    "reader2 = reader2.reset_index(drop=True)\n",
    "reader2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_col = ['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA', 'start', 'end', 'type', 'distClosestG4', 'CGI', 'Ori', 'TSS']\n",
    "col_G4 = ['start', 'end', 'type', 'distClosestG4', 'CGI', 'Ori', 'TSS']\n",
    "table_finale = pd.DataFrame(columns = my_col)\n",
    "table_finale[['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA']] = reader2\n",
    "\n",
    "G4_table = reader[reader['chr'] == 'chr'+str(i_chrom)]\n",
    "G4_table = G4_table.reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_finale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "G4_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_mut = len(reader2)\n",
    "n_G4 = len(G4_table)\n",
    "i_G4 = 0\n",
    "#start_table = G4_table['start']\n",
    "end_table = G4_table['end']\n",
    "pos_table = reader2['pos']\n",
    "\n",
    "for i_mut in range (n_mut):\n",
    "    #start = start_table.loc[i_G4]\n",
    "    end = end_table.loc[i_G4]\n",
    "    pos = pos_table.loc[i_mut]\n",
    "    while pos > end :\n",
    "        i_G4 += 1\n",
    "        end = end_table.loc[i_G4]\n",
    "    table_finale.loc[i_mut,col_G4] = G4_table.loc[i_G4]\n",
    "    if i_mut % 1000 == 0:\n",
    "        print(i_mut)\n",
    "print(\"fini\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_finale.to_csv(f'chr_{i_chrom}_final.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Automatique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>distClosestG4</th>\n",
       "      <th>CGI</th>\n",
       "      <th>Ori</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11008</td>\n",
       "      <td>11026</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>gggcgggggttggggggg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11058</td>\n",
       "      <td>11078</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>gggctggggcggggggaggg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12487</td>\n",
       "      <td>12514</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>1410</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGTGGTAGGTGCAGAGACGGGAGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14664</td>\n",
       "      <td>14683</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>1267</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGTCTGGGGGGGAAGGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15949</td>\n",
       "      <td>15973</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>687</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGAGCCGGGAGGTGGGGAACAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732447</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59362173</td>\n",
       "      <td>59362191</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>370</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>gggcgggcgggctgaggg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732448</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59362560</td>\n",
       "      <td>59362587</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>gggtgagggtgagggttggggttgggg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732449</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59362595</td>\n",
       "      <td>59363470</td>\n",
       "      <td>G-track</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ggggttggggttggggttggggttggggttagggtttgggtttagg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732450</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363480</td>\n",
       "      <td>59363508</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>gggtgtgtgtgggtgtggtgtgtgtggg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732451</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363518</td>\n",
       "      <td>59363547</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>gggtgtgggtgtgggtgtgggtgtgtggg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>732452 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr     start       end       type  distClosestG4  CGI  Ori  TSS  \\\n",
       "0       chr1     11008     11026  Canonical             33    0    0    0   \n",
       "1       chr1     11058     11078  Canonical             33    0    0    0   \n",
       "2       chr1     12487     12514      Bulge           1410    0    0    0   \n",
       "3       chr1     14664     14683      Bulge           1267    0    0    0   \n",
       "4       chr1     15949     15973      Bulge            687    0    0    0   \n",
       "...      ...       ...       ...        ...            ...  ...  ...  ...   \n",
       "732447  chrY  59362173  59362191  Canonical            370    0    0    0   \n",
       "732448  chrY  59362560  59362587  Canonical              9    0    0    0   \n",
       "732449  chrY  59362595  59363470    G-track              9    0    0    0   \n",
       "732450  chrY  59363480  59363508      Bulge             11    0    0    0   \n",
       "732451  chrY  59363518  59363547  Canonical             11    0    0    0   \n",
       "\n",
       "                                                 sequence  \n",
       "0                                      gggcgggggttggggggg  \n",
       "1                                    gggctggggcggggggaggg  \n",
       "2                             GGGTGGTAGGTGCAGAGACGGGAGGGG  \n",
       "3                                     GGGTCTGGGGGGGAAGGTG  \n",
       "4                                GGAGCCGGGAGGTGGGGAACAGGG  \n",
       "...                                                   ...  \n",
       "732447                                 gggcgggcgggctgaggg  \n",
       "732448                        gggtgagggtgagggttggggttgggg  \n",
       "732449  ggggttggggttggggttggggttggggttagggtttgggtttagg...  \n",
       "732450                       gggtgtgtgtgggtgtggtgtgtgtggg  \n",
       "732451                      gggtgtgggtgtgggtgtgggtgtgtggg  \n",
       "\n",
       "[732452 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = open('C:/Users/BlueC/Documents/Mines/Santé/Code/annotated_G4plus_categories_explicit.csv')\n",
    "reader = pd.read_csv(file)\n",
    "reader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr 1\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "fini\n",
      "chr 2\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "fini\n",
      "chr 3\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "fini\n",
      "chr 4\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 5\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 6\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 7\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "fini\n",
      "chr 8\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 9\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 10\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 11\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "fini\n",
      "chr 12\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 13\n",
      "0\n",
      "10000\n",
      "fini\n",
      "chr 14\n",
      "0\n",
      "10000\n",
      "20000\n",
      "fini\n",
      "chr 15\n",
      "0\n",
      "10000\n",
      "20000\n",
      "fini\n",
      "chr 16\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 17\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "fini\n",
      "chr 18\n",
      "0\n",
      "10000\n",
      "fini\n",
      "chr 19\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "fini\n",
      "chr 20\n",
      "0\n",
      "10000\n",
      "20000\n",
      "fini\n",
      "chr 21\n",
      "0\n",
      "10000\n",
      "fini\n",
      "chr 22\n",
      "0\n",
      "10000\n",
      "20000\n",
      "fini\n",
      "chr X\n",
      "0\n",
      "10000\n",
      "20000\n",
      "fini\n",
      "chr Y\n",
      "0\n",
      "fini\n"
     ]
    }
   ],
   "source": [
    "liste = [i for i in range (1,23)] + ['X','Y']\n",
    "for i_chrom in (liste):\n",
    "    print(\"chr\",i_chrom)\n",
    "    file2 = open(f'C:/Users/BlueC/Documents/Mines/Santé/Code/chr/chr_{i_chrom}_2')\n",
    "    my_col = ['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA']\n",
    "    separateurs = \"\\t|AA=\"\n",
    "    reader2 = pd.read_csv(file2, sep = separateurs, names = my_col, header = None, engine ='python')\n",
    "    reader2 = reader2.drop_duplicates()\n",
    "    reader2 = reader2.reset_index(drop=True)\n",
    "    \n",
    "    my_col = ['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA', 'start', 'end', 'type', 'distClosestG4', 'CGI', 'Ori', 'TSS','sequence']\n",
    "    col_G4 = ['start', 'end', 'type', 'distClosestG4', 'CGI', 'Ori', 'TSS','sequence']\n",
    "    table_finale = pd.DataFrame(columns = my_col)\n",
    "    table_finale[['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA']] = reader2\n",
    "\n",
    "    G4_table = reader[reader['chr'] == 'chr'+str(i_chrom)]\n",
    "    G4_table = G4_table.reset_index(drop = True)\n",
    "    \n",
    "    n_mut = len(reader2)\n",
    "    n_G4 = len(G4_table)\n",
    "    i_G4 = 0\n",
    "    #start_table = G4_table['start']\n",
    "    end_table = G4_table['end']\n",
    "    pos_table = reader2['pos']\n",
    "\n",
    "    for i_mut in range (n_mut):\n",
    "        #start = start_table.loc[i_G4]\n",
    "        end = end_table.loc[i_G4]\n",
    "        pos = pos_table.loc[i_mut]\n",
    "        while pos > end :\n",
    "            i_G4 += 1\n",
    "            end = end_table.loc[i_G4]\n",
    "        table_finale.loc[i_mut,col_G4] = G4_table.loc[i_G4]\n",
    "        if i_mut % 10000 == 0:\n",
    "            print(i_mut)\n",
    "    print(\"fini\")\n",
    "    table_finale = table_finale[table_finale['pos']>=table_finale['start']]\n",
    "    table_finale.to_csv(f'C:/Users/BlueC/Documents/Mines/Santé/Code/chr/chr_{i_chrom}_final.csv', index = False)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Création des datasets de mutations - positions random"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Création des fichiers de comparaison - fichiers sans N"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A ne pas refaire tourner "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open('C:/Users/BlueC/Documents/Mines/Santé/Code/random_G4_explicit.csv')\n",
    "reader = pd.read_csv(file, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>distClosestG4</th>\n",
       "      <th>CGI</th>\n",
       "      <th>Ori</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2402</td>\n",
       "      <td>2426</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>473</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2899</td>\n",
       "      <td>2933</td>\n",
       "      <td>LongLoop</td>\n",
       "      <td>473</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10525</td>\n",
       "      <td>10551</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>918</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gccttcagagtaccaccgaaatctgt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11469</td>\n",
       "      <td>11503</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>918</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGGAAACACCCGGAGCATATGCTGTTTGGTCTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16111</td>\n",
       "      <td>16130</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>2181</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGGCACTGACTTCTGCAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720229</th>\n",
       "      <td>chr21</td>\n",
       "      <td>48113314</td>\n",
       "      <td>48113335</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>4140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gtctttgccccattttatatt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720230</th>\n",
       "      <td>chr21</td>\n",
       "      <td>48117475</td>\n",
       "      <td>48117497</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>2462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>atcatcattcactctgcaataa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720231</th>\n",
       "      <td>chr21</td>\n",
       "      <td>48119959</td>\n",
       "      <td>48119981</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>2462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720232</th>\n",
       "      <td>chr21</td>\n",
       "      <td>48123506</td>\n",
       "      <td>48123534</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>2176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720233</th>\n",
       "      <td>chr21</td>\n",
       "      <td>48125710</td>\n",
       "      <td>48125740</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>2176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>720234 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr     start       end       type  distClosestG4  CGI  Ori  TSS  \\\n",
       "0        chr1      2402      2426      Bulge            473  NaN  NaN  NaN   \n",
       "1        chr1      2899      2933   LongLoop            473  NaN  NaN  NaN   \n",
       "2        chr1     10525     10551      Bulge            918  NaN  NaN  NaN   \n",
       "3        chr1     11469     11503  Canonical            918  NaN  NaN  NaN   \n",
       "4        chr1     16111     16130      Bulge           2181  NaN  NaN  NaN   \n",
       "...       ...       ...       ...        ...            ...  ...  ...  ...   \n",
       "720229  chr21  48113314  48113335       GVBQ           4140  NaN  NaN  NaN   \n",
       "720230  chr21  48117475  48117497  Canonical           2462  NaN  NaN  NaN   \n",
       "720231  chr21  48119959  48119981       GVBQ           2462  NaN  NaN  NaN   \n",
       "720232  chr21  48123506  48123534      Bulge           2176  NaN  NaN  NaN   \n",
       "720233  chr21  48125710  48125740       GVBQ           2176  NaN  NaN  NaN   \n",
       "\n",
       "                                  sequence  \n",
       "0                 NNNNNNNNNNNNNNNNNNNNNNNN  \n",
       "1       NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN  \n",
       "2               gccttcagagtaccaccgaaatctgt  \n",
       "3       GGGAAACACCCGGAGCATATGCTGTTTGGTCTCA  \n",
       "4                      GGGCACTGACTTCTGCAAC  \n",
       "...                                    ...  \n",
       "720229               gtctttgccccattttatatt  \n",
       "720230              atcatcattcactctgcaataa  \n",
       "720231              NNNNNNNNNNNNNNNNNNNNNN  \n",
       "720232        NNNNNNNNNNNNNNNNNNNNNNNNNNNN  \n",
       "720233      NNNNNNNNNNNNNNNNNNNNNNNNNNNNNN  \n",
       "\n",
       "[720234 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On observe des séquences comportant des N, que l'on ne veut pas considérer. On cherche donc à supprimer ces lignes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "contain_values = reader[~reader['sequence'].str.contains('N')]\n",
    "contain_values = contain_values.reset_index(drop=True)\n",
    "contain_values.to_csv('C:/Users/BlueC/Documents/Mines/Santé/Code/random_G4_explicit_sansN.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A ne pas refaire tourner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open('C:/Users/BlueC/Documents/Mines/Santé/Code/random_G4_explicit_sansN.csv')\n",
    "reader = pd.read_csv(file, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range (1,23):\n",
    "    test = reader[(reader['chr'] == f'chr{i}')]\n",
    "    test = test[['chr','start','end']]\n",
    "    test = test.sort_values(['start','end'])\n",
    "    test['chr'] = i\n",
    "    path = f\"c:/Users/BlueC/Documents/Mines/Santé/Code/np_random_{i}_sansN.txt\"\n",
    "    np.savetxt(path, test.values, fmt='%s', delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in (['X','Y']):\n",
    "    test = reader[(reader['chr'] == f'chr{i}')]\n",
    "    test = test[['chr','start','end']]\n",
    "    test = test.sort_values(['start','end'])\n",
    "    test['chr'] = i\n",
    "    path = f\"c:/Users/BlueC/Documents/Mines/Santé/Code/np_random_{i}_sansN.txt\"\n",
    "    np.savetxt(path, test.values, fmt='%s', delimiter='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Création de la base recensant toutes les mutations ayant eu lieu sur les positions aléatoires"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On récupère les fichiers générés grâce au code bash précédemment évoqué pour créer une base de données complète, recensant les mutations et les caractéristiques des positions auxquelles elles sont associées."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open('C:/Users/BlueC/Documents/Mines/Santé/Code/random_G4_explicit.csv')\n",
    "reader = pd.read_csv(file, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i_chrom = '6'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file2 = open(f'C:/Users/BlueC/Documents/Mines/Santé/Code/chr_random/chr_random_{i_chrom}_sansN')\n",
    "my_col = ['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA']\n",
    "separateurs = \"\\t|AA=\"\n",
    "reader2 = pd.read_csv(file2, sep = separateurs, names = my_col, header = None, engine ='python')\n",
    "reader2 = reader2.drop_duplicates()\n",
    "reader2 = reader2.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_col = ['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA', 'start', 'end', 'type', 'distClosestG4', 'CGI', 'Ori', 'TSS', 'sequence']\n",
    "col_G4 = ['start', 'end', 'type', 'distClosestG4', 'CGI', 'Ori', 'TSS', 'sequence']\n",
    "table_finale = pd.DataFrame(columns = my_col)\n",
    "table_finale[['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA']] = reader2\n",
    "\n",
    "G4_table = reader[reader['chr'] == 'chr'+str(i_chrom)]\n",
    "G4_table = G4_table.reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_finale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "G4_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_mut = len(reader2)\n",
    "n_G4 = len(G4_table)\n",
    "i_G4 = 0\n",
    "#start_table = G4_table['start']\n",
    "end_table = G4_table['end']\n",
    "pos_table = reader2['pos']\n",
    "\n",
    "for i_mut in range (n_mut):\n",
    "    #start = start_table.loc[i_G4]\n",
    "    end = end_table.loc[i_G4]\n",
    "    pos = pos_table.loc[i_mut]\n",
    "    while pos > end :\n",
    "        i_G4 += 1\n",
    "        end = end_table.loc[i_G4]\n",
    "    table_finale.loc[i_mut,col_G4] = G4_table.loc[i_G4]\n",
    "    if i_mut % 1000 == 0:\n",
    "        print(i_mut)\n",
    "print(\"fini\")\n",
    "table_finale = table_finale[table_finale['pos']>=table_finale['start']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_finale = table_finale[table_finale['pos']>=table_finale['start']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_finale[table_finale['sequence'].str.contains('N')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_finale.to_csv(f'chr_random_{i_chrom}_final_sansN.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Automatique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "liste = [i for i in range (17,23)] + ['X','Y']\n",
    "for i_chrom in (liste):\n",
    "    print(\"chr\",i_chrom)\n",
    "    file2 = open(f'C:/Users/BlueC/Documents/Mines/Santé/Code/chr_random/chr_random_{i_chrom}_sansN')\n",
    "    my_col = ['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA']\n",
    "    separateurs = \"\\t|AA=\"\n",
    "    reader2 = pd.read_csv(file2, sep = separateurs, names = my_col, header = None, engine ='python')\n",
    "    reader2 = reader2.drop_duplicates()\n",
    "    reader2 = reader2.reset_index(drop=True)\n",
    "    \n",
    "    my_col = ['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA', 'start', 'end', 'type', 'distClosestG4', 'CGI', 'Ori', 'TSS', 'sequence']\n",
    "    col_G4 = ['start', 'end', 'type', 'distClosestG4', 'CGI', 'Ori', 'TSS', 'sequence']\n",
    "    table_finale = pd.DataFrame(columns = my_col)\n",
    "    table_finale[['chr', 'pos', 'name', 'orig', 'mut', '%', 'PASS','INFOS', 'AA']] = reader2\n",
    "\n",
    "    G4_table = reader[reader['chr'] == 'chr'+str(i_chrom)]\n",
    "    G4_table = G4_table.reset_index(drop = True)\n",
    "    \n",
    "    n_mut = len(reader2)\n",
    "    n_G4 = len(G4_table)\n",
    "    i_G4 = 0\n",
    "    #start_table = G4_table['start']\n",
    "    end_table = G4_table['end']\n",
    "    pos_table = reader2['pos']\n",
    "\n",
    "    for i_mut in range (n_mut):\n",
    "        #start = start_table.loc[i_G4]\n",
    "        end = end_table.loc[i_G4]\n",
    "        pos = pos_table.loc[i_mut]\n",
    "        while pos > end :\n",
    "            i_G4 += 1\n",
    "            end = end_table.loc[i_G4]\n",
    "        table_finale.loc[i_mut,col_G4] = G4_table.loc[i_G4]\n",
    "        if i_mut % 10000 == 0:\n",
    "            print(i_mut)\n",
    "    print(\"fini\")\n",
    "    \n",
    "    table_finale = table_finale[table_finale['pos']>=table_finale['start']]\n",
    "    table_finale.to_csv(f'chr_random_{i_chrom}_final_sansN.csv', index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Vérification des fichiers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open('C:/Users/BlueC/Documents/Mines/Santé/Code/random_G4_explicit.csv')\n",
    "reader = pd.read_csv(file, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "reader_chr2 = reader[reader['chr'] == 'chr2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>distClosestG4</th>\n",
       "      <th>CGI</th>\n",
       "      <th>Ori</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64261</th>\n",
       "      <td>chr2</td>\n",
       "      <td>3882</td>\n",
       "      <td>3905</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64262</th>\n",
       "      <td>chr2</td>\n",
       "      <td>4110</td>\n",
       "      <td>4132</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64263</th>\n",
       "      <td>chr2</td>\n",
       "      <td>6065</td>\n",
       "      <td>6087</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>1268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64264</th>\n",
       "      <td>chr2</td>\n",
       "      <td>7355</td>\n",
       "      <td>7378</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>308</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64265</th>\n",
       "      <td>chr2</td>\n",
       "      <td>7686</td>\n",
       "      <td>7710</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>308</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116891</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243132031</td>\n",
       "      <td>243132047</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116892</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243132493</td>\n",
       "      <td>243132517</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>446</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116893</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243133964</td>\n",
       "      <td>243133991</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>1447</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116894</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243148070</td>\n",
       "      <td>243148095</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>2027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116895</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243150122</td>\n",
       "      <td>243150142</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>2027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1085 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end       type  distClosestG4  CGI  Ori  TSS  \\\n",
       "64261   chr2       3882       3905  Canonical            205  NaN  NaN  NaN   \n",
       "64262   chr2       4110       4132  Canonical            205  NaN  NaN  NaN   \n",
       "64263   chr2       6065       6087      Bulge           1268  NaN  NaN  NaN   \n",
       "64264   chr2       7355       7378       GVBQ            308  NaN  NaN  NaN   \n",
       "64265   chr2       7686       7710       GVBQ            308  NaN  NaN  NaN   \n",
       "...      ...        ...        ...        ...            ...  ...  ...  ...   \n",
       "116891  chr2  243132031  243132047      Bulge            330  NaN  NaN  NaN   \n",
       "116892  chr2  243132493  243132517      Bulge            446  NaN  NaN  NaN   \n",
       "116893  chr2  243133964  243133991       GVBQ           1447  NaN  NaN  NaN   \n",
       "116894  chr2  243148070  243148095      Bulge           2027  NaN  NaN  NaN   \n",
       "116895  chr2  243150122  243150142  Canonical           2027  NaN  NaN  NaN   \n",
       "\n",
       "                           sequence  \n",
       "64261       NNNNNNNNNNNNNNNNNNNNNNN  \n",
       "64262        NNNNNNNNNNNNNNNNNNNNNN  \n",
       "64263        NNNNNNNNNNNNNNNNNNNNNN  \n",
       "64264       NNNNNNNNNNNNNNNNNNNNNNN  \n",
       "64265      NNNNNNNNNNNNNNNNNNNNNNNN  \n",
       "...                             ...  \n",
       "116891             NNNNNNNNNNNNNNNN  \n",
       "116892     NNNNNNNNNNNNNNNNNNNNNNNN  \n",
       "116893  NNNNNNNNNNNNNNNNNNNNNNNNNNN  \n",
       "116894    NNNNNNNNNNNNNNNNNNNNNNNNN  \n",
       "116895         NNNNNNNNNNNNNNNNNNNN  \n",
       "\n",
       "[1085 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reader_chr2[reader_chr2['sequence'].str.contains('N')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "file2 = open('C:/Users/BlueC/Documents/Mines/Santé/Code/random_G4_explicit_sansN.csv')\n",
    "reader2 = pd.read_csv(file2, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>distClosestG4</th>\n",
       "      <th>CGI</th>\n",
       "      <th>Ori</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57985</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15937</td>\n",
       "      <td>15965</td>\n",
       "      <td>Canonical</td>\n",
       "      <td>7007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCTGGCTCTATTTTGGGTGACATGCTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57986</th>\n",
       "      <td>chr2</td>\n",
       "      <td>39121</td>\n",
       "      <td>39146</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>2013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGCCGTCACTATTGCCCATTCTTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57987</th>\n",
       "      <td>chr2</td>\n",
       "      <td>41159</td>\n",
       "      <td>41188</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>2013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TGATTACTTTGATTTAAATGATGTTAATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57988</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50770</td>\n",
       "      <td>50791</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcccaggaatgtattctaagg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57989</th>\n",
       "      <td>chr2</td>\n",
       "      <td>51038</td>\n",
       "      <td>51061</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gggggctcctgatatctttttgg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109539</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243177646</td>\n",
       "      <td>243177662</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ggttgtctattggctt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109540</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243179828</td>\n",
       "      <td>243179850</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAGAAGAGAGCATGGTTACATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109541</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243180374</td>\n",
       "      <td>243180401</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAAGAATCAACCTGGGAGCCAAGCCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109542</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243180811</td>\n",
       "      <td>243180834</td>\n",
       "      <td>Bulge</td>\n",
       "      <td>410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TATGATTCCCTTGCTGCCTGGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109543</th>\n",
       "      <td>chr2</td>\n",
       "      <td>243187703</td>\n",
       "      <td>243187728</td>\n",
       "      <td>GVBQ</td>\n",
       "      <td>6869</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>acccaggaggtggaggttgcaggga</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51559 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end       type  distClosestG4  CGI  Ori  TSS  \\\n",
       "57985   chr2      15937      15965  Canonical           7007  NaN  NaN  NaN   \n",
       "57986   chr2      39121      39146      Bulge           2013  NaN  NaN  NaN   \n",
       "57987   chr2      41159      41188       GVBQ           2013  NaN  NaN  NaN   \n",
       "57988   chr2      50770      50791       GVBQ            247  NaN  NaN  NaN   \n",
       "57989   chr2      51038      51061      Bulge            247  NaN  NaN  NaN   \n",
       "...      ...        ...        ...        ...            ...  ...  ...  ...   \n",
       "109539  chr2  243177646  243177662       GVBQ             65  NaN  NaN  NaN   \n",
       "109540  chr2  243179828  243179850      Bulge            524  NaN  NaN  NaN   \n",
       "109541  chr2  243180374  243180401      Bulge            410  NaN  NaN  NaN   \n",
       "109542  chr2  243180811  243180834      Bulge            410  NaN  NaN  NaN   \n",
       "109543  chr2  243187703  243187728       GVBQ           6869  NaN  NaN  NaN   \n",
       "\n",
       "                             sequence  \n",
       "57985    TCTGGCTCTATTTTGGGTGACATGCTCC  \n",
       "57986       AGCCGTCACTATTGCCCATTCTTGT  \n",
       "57987   TGATTACTTTGATTTAAATGATGTTAATC  \n",
       "57988           tcccaggaatgtattctaagg  \n",
       "57989         gggggctcctgatatctttttgg  \n",
       "...                               ...  \n",
       "109539               ggttgtctattggctt  \n",
       "109540         AAGAAGAGAGCATGGTTACATT  \n",
       "109541    AAAGAATCAACCTGGGAGCCAAGCCCA  \n",
       "109542        TATGATTCCCTTGCTGCCTGGAA  \n",
       "109543      acccaggaggtggaggttgcaggga  \n",
       "\n",
       "[51559 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reader2_chr2 = reader2[reader2['chr'] == 'chr2']\n",
    "reader2_chr2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>distClosestG4</th>\n",
       "      <th>CGI</th>\n",
       "      <th>Ori</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chr, start, end, type, distClosestG4, CGI, Ori, TSS, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reader2_chr2[reader2_chr2['sequence'].str.contains('N')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Il n'y a plus de N dans la base"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
